File: client-sdk/typescript/src/subjects.ts
// Subject construction and validation per protocol §2.
//
// Wire layout (v0.3): `agents.{verb}.{agent}.{owner}.{name}` where `verb`
// is one of the protocol-reserved verbs `prompt` / `hb` (heartbeat,
// abbreviated for wire economy) / `status` (plus `attachments` reserved
// for the future §5.5 endpoint). The 5th token is the instance name.
//
// Token validation (assertValidToken / isRecommendedToken) is enforced
// only at MUST-rule level — anything NATS-safe is accepted. Callers that
// want stricter SHOULD-rule conformance use `isRecommendedToken`.
//
// `AgentSubject.new(...)` is the canonical entry point: validates the
// three identifying tokens and exposes `.prompt`, `.heartbeat`, `.status`
// getters that build the v0.3 verb-first subjects. Centralised so the
// verb-first wire shape lives in exactly one place across the SDK,
// agents, and examples.

const FORBIDDEN_CHARS = /[.*>\s\0]/;
const RECOMMENDED_CHARS = /^[a-z0-9_-]+$/;

export class InvalidSubjectTokenError extends Error {
  constructor(
    public readonly role: string,
    public readonly token: string,
    reason: string,
  ) {
    super(`invalid ${role} token "${token}": ${reason}`);
    this.name = "InvalidSubjectTokenError";
  }
}

/** Throws {@link InvalidSubjectTokenError} when the token violates MUST rules. */
export function assertValidToken(token: string, role: string): void {
  if (token.length === 0) {
    throw new InvalidSubjectTokenError(role, token, "must not be empty");
  }
  if (token.startsWith("$")) {
    throw new InvalidSubjectTokenError(role, token, "must not begin with '$'");
  }
  if (FORBIDDEN_CHARS.test(token)) {
    throw new InvalidSubjectTokenError(
      role,
      token,
      "must not contain whitespace, '.', '*', '>', or NUL",
    );
  }
}

/** @returns true iff the token is within the spec's SHOULD character set. */
export function isRecommendedToken(token: string): boolean {
  return token.length >= 1 && token.length <= 63 && RECOMMENDED_CHARS.test(token);
}

/** Subject root (§2). */
export const SUBJECT_ROOT = "agents";

// §2 (v0.3) reserved verbs. `prompt` / `hb` / `status` are wired up by
// this SDK; `attachments` is reserved for the future §5.5 endpoint. The
// heartbeat verb is the abbreviation `hb` on the wire (§8.1) — kept
// short because heartbeat traffic dominates per-account subject volume.
export const VERB_PROMPT = "prompt";
export const VERB_HEARTBEAT = "hb";
export const VERB_STATUS = "status";
export const VERB_ATTACHMENTS = "attachments";

export const RESERVED_VERBS: ReadonlySet<string> = new Set([
  VERB_PROMPT,
  VERB_HEARTBEAT,
  VERB_STATUS,
  VERB_ATTACHMENTS,
]);

/** Number of tokens in a v0.3 agent subject: `agents.{verb}.{agent}.{owner}.{name}`. */
const SUBJECT_TOKEN_COUNT = 5;

/**
 * Optional construction knobs for {@link AgentSubject.new}.
 */
export interface AgentSubjectOptions {
  /**
   * Override the wire token used in the subject's 3rd position. Defaults
   * to `agent` (i.e. `metadata.agent` and the subject token are the same).
   *
   * Allows callers to advertise a long, canonical name in `metadata.agent`
   * (e.g. `"claude-code"`, `"openclaw"`) while keeping the subject's 3rd
   * token short for wire-economy reasons (e.g. `"cc"`, `"oc"`). Both
   * tokens are validated against §2 MUST rules independently.
   */
  readonly subjectToken?: string;
}

/**
 * The three identifying tokens of an agent, validated against §2 MUST rules.
 *
 * Construct via {@link AgentSubject.new} — direct construction bypasses
 * validation and SHOULD NOT be used by callers.
 */
export class AgentSubject {
  private constructor(
    public readonly agent: string,
    public readonly owner: string,
    public readonly name: string,
    public readonly subjectToken: string,
  ) {}

  /**
   * Validate the identifying tokens and return an {@link AgentSubject}.
   *
   * `agent` carries the canonical agent identifier (the value that goes
   * into `$SRV.INFO` `metadata.agent`). `opts.subjectToken`, if given,
   * is the wire token used in the subject's 3rd position; when omitted,
   * the subject token equals `agent`.
   *
   * Throws {@link InvalidSubjectTokenError} when any token violates §2
   * MUST rules.
   */
  static new(
    agent: string,
    owner: string,
    name: string,
    opts: AgentSubjectOptions = {},
  ): AgentSubject {
    assertValidToken(agent, "agent");
    assertValidToken(owner, "owner");
    assertValidToken(name, "name");
    const subjectToken = opts.subjectToken ?? agent;
    if (subjectToken !== agent) {
      assertValidToken(subjectToken, "subjectToken");
    }
    return new AgentSubject(agent, owner, name, subjectToken);
  }

  /** The agent's prompt subject — `agents.prompt.{subjectToken}.{owner}.{name}` (§2 v0.3). */
  get prompt(): string {
    return `${SUBJECT_ROOT}.${VERB_PROMPT}.${this.subjectToken}.${this.owner}.${this.name}`;
  }

  /** The agent's heartbeat subject — `agents.hb.{subjectToken}.{owner}.{name}` (§8.1 v0.3). */
  get heartbeat(): string {
    return `${SUBJECT_ROOT}.${VERB_HEARTBEAT}.${this.subjectToken}.${this.owner}.${this.name}`;
  }

  /** The agent's status request/response subject — `agents.status.{subjectToken}.{owner}.{name}` (§8.7 (v0.3)). */
  get status(): string {
    return `${SUBJECT_ROOT}.${VERB_STATUS}.${this.subjectToken}.${this.owner}.${this.name}`;
  }
}

/** True iff `subject` matches `agents.hb.{agent}.{owner}.{name}` (§8.1 v0.3). */
export function isHeartbeatSubject(subject: string): boolean {
  const parts = subject.split(".");
  return (
    parts.length === SUBJECT_TOKEN_COUNT && parts[0] === SUBJECT_ROOT && parts[1] === VERB_HEARTBEAT
  );
}

export interface ParseAgentSubjectOptions {
  /** Verb to require in token 2. Defaults to `VERB_PROMPT`. */
  readonly verb?: string;
}

/**
 * Parse an `agents.{verb}.{agent}.{owner}.{name}` subject into an
 * {@link AgentSubject}.
 *
 * Returns `null` when the subject has the wrong root, is the wrong verb
 * (default `prompt`), or fails token validation. Pass a different `verb`
 * to parse heartbeat / status subjects through the same helper.
 *
 * **Note on the canonical agent identifier.** The 3rd subject token is
 * the *wire token* — for harnesses that set `subjectToken` to an
 * abbreviation (e.g. `"cc"` for `claude-code`, `"oc"` for `openclaw`),
 * `AgentSubject.agent` returned here is the abbreviation, **not** the
 * canonical `metadata.agent` value. The canonical name only lives in
 * `$SRV.INFO.metadata.agent`; callers that need it should use
 * `Agents.lookupInstance(instanceId)` (or `Agents.discover()`) and read
 * `AgentInfo.agent`. Filtering on `subject.agent` matches the wire token,
 * filtering on `AgentInfo.agent` matches the canonical identifier.
 */
export function parseAgentSubject(
  subject: string,
  opts: ParseAgentSubjectOptions = {},
): AgentSubject | null {
  const verb = opts.verb ?? VERB_PROMPT;
  const parts = subject.split(".");
  if (parts.length !== SUBJECT_TOKEN_COUNT || parts[0] !== SUBJECT_ROOT || parts[1] !== verb) {
    return null;
  }
  const [, , agent, owner, name] = parts;
  if (agent === undefined || owner === undefined || name === undefined) return null;
  try {
    return AgentSubject.new(agent, owner, name);
  } catch {
    return null;
  }
}

File: client-sdk/python/src/synadia_ai/agents/subjects.py
"""Subject construction and validation per protocol §2.

Wire layout (v0.3): ``agents.{verb}.{agent}.{owner}.{name}`` where ``verb``
is one of the protocol-reserved verbs ``prompt``, ``heartbeat``, ``status``
(plus ``attachments`` reserved for the future §5.5 endpoint). Verbs and
instance names live in different positions, so an agent literally named
``heartbeat`` no longer collides with the §8 heartbeat subject.

``agent``, ``owner``, and ``name`` are constrained to lowercase alphanumeric
plus hyphens (``agent``) or hyphens and underscores (``owner``, ``name``).
Tokens that would otherwise be invalid NATS subject characters are escaped
internally as base64-url-no-padding — the SDK's "sanitize sensibly"
implementation detail, not a protocol contract.

§2 uses ``agent`` where v0.0.1 of this SDK used ``platform``; the rename
is tracked in CHANGELOG.md under 0.1.0. The verb-first move landed in
0.4.0 (protocol v0.3).
"""

from __future__ import annotations

import base64
import re
from dataclasses import dataclass

from .errors import InvalidSubjectToken

ROOT = "agents"

# §2 token constraints.
_AGENT_RE = re.compile(r"^[a-z0-9-]+$")
_OWNER_RE = re.compile(r"^[a-z0-9_-]+$")
_NAME_RE = _OWNER_RE

# §2 (v0.3) reserved verbs. ``prompt``/``heartbeat``/``status`` are wired up
# by this SDK; ``attachments`` is reserved for the future §5.5 endpoint.
VERB_PROMPT = "prompt"
VERB_HEARTBEAT = "heartbeat"
VERB_STATUS = "status"
VERB_ATTACHMENTS = "attachments"

RESERVED_VERBS = frozenset({VERB_PROMPT, VERB_HEARTBEAT, VERB_STATUS, VERB_ATTACHMENTS})

# `agents.{verb}.{agent}.{owner}.{name}` — 5 tokens (§2 v0.3).
_SUBJECT_TOKEN_COUNT = 5


@dataclass(frozen=True, slots=True)
class AgentSubject:
    """The three identifying tokens of an agent, already validated/sanitized.

    Construct via :meth:`new` — it enforces §1 constraints and falls back to
    SDK-internal base64 escaping for owner/name tokens that contain characters
    NATS subjects can't carry directly.
    """

    agent: str
    owner: str
    name: str

    @classmethod
    def new(cls, agent: str, owner: str, name: str) -> AgentSubject:
        if not agent:
            raise InvalidSubjectToken("agent must be non-empty")
        if not _AGENT_RE.fullmatch(agent):
            raise InvalidSubjectToken(
                f"agent {agent!r} must be lowercase alphanumeric + hyphens (§2)"
            )
        return cls(
            agent=agent,
            owner=_sanitize(owner, "owner", _OWNER_RE),
            name=_sanitize(name, "name", _NAME_RE),
        )

    @property
    def inbox(self) -> str:
        """Backwards-name-compat alias of :attr:`prompt`."""
        return self.prompt

    @property
    def prompt(self) -> str:
        """The agent's prompt subject (§2)."""
        return f"{ROOT}.{VERB_PROMPT}.{self.agent}.{self.owner}.{self.name}"

    @property
    def heartbeat(self) -> str:
        """The agent's heartbeat subject (§8.1)."""
        return f"{ROOT}.{VERB_HEARTBEAT}.{self.agent}.{self.owner}.{self.name}"

    @property
    def status(self) -> str:
        """The agent's status request/response subject (v0.3 §-TBD)."""
        return f"{ROOT}.{VERB_STATUS}.{self.agent}.{self.owner}.{self.name}"


def _sanitize(token: str, field: str, pattern: re.Pattern[str]) -> str:
    """Validate a token or escape it as base64-url-no-padding when necessary.

    Natural tokens (matching the strict lowercase `pattern`) pass through
    verbatim so subjects stay readable for the common case. Non-conforming
    tokens are base64-url-no-padding encoded — the encoding alphabet
    `[A-Za-z0-9_-]` is always valid inside a NATS subject token, so the
    escape form is guaranteed safe even though it breaks the natural-token
    lowercase convention. This is an SDK implementation detail ("sanitize
    sensibly") — the protocol does not mandate the encoding scheme.
    """
    if not token:
        raise InvalidSubjectToken(f"{field} must be non-empty")
    if pattern.fullmatch(token):
        return token
    return base64.urlsafe_b64encode(token.encode("utf-8")).rstrip(b"=").decode("ascii")


def is_heartbeat_subject(subject: str) -> bool:
    """True iff the subject is of the form `agents.heartbeat.{a}.{o}.{n}`."""
    parts = subject.split(".")
    return len(parts) == _SUBJECT_TOKEN_COUNT and parts[0] == ROOT and parts[1] == VERB_HEARTBEAT


def parse_agent_subject(subject: str, *, verb: str = VERB_PROMPT) -> AgentSubject | None:
    """Parse an `agents.{verb}.{agent}.{owner}.{name}` subject.

    Returns ``None`` when the subject is not the expected verb (default
    ``prompt``), has the wrong root, or fails token validation. Pass a
    different ``verb`` to parse heartbeat / status subjects through the
    same helper.
    """
    parts = subject.split(".")
    if len(parts) != _SUBJECT_TOKEN_COUNT or parts[0] != ROOT or parts[1] != verb:
        return None
    _, _, agent, owner, name = parts
    try:
        return AgentSubject.new(agent=agent, owner=owner, name=name)
    except InvalidSubjectToken:
        return None
